### PR TITLE
Cli/parse with linked lists

### PR DESCRIPTION
**Pull request by Beza20**
With this pull request we have enabled a user to input multiple commands at once using a connecting word "and." This was done by using a helper function <strtokstr_r> that has very similar structure as the C built-in function <strtok_r.> It parses through a string and tokenizes a segment of that string whenever it finds the inputed delimiter of a string type. Then we use the helper function to read the cmd line and store the tokenized segment of the input into a utlist. Note: Each element in the utlist has to further be parsed breaking it up using a space. And then in the UI module, we made sure that all of the command segments put into the utlist are iteratively changed into a command struct by calling the <cmd_from_string> and executed. And with that this pull request will introduce the "and" feature to chiventure resolving issue #275.

### CODE REVIEW

File: src/cli/src/parser.c [hunk 1]
```diff
@@ -3,7 +3,95 @@
...
+    if (s == NULL)
+    {
+        s = *save_ptr;
+    }
+    else if (s == NULL || *s == '\0')
+    {
+        *save_ptr = s;
+        return NULL;
+    }
+    while (strstr(s,delim)==s)
```
Comment (ecoble) [inline src/cli/src/parser.c:30]: Please put spaces around ==

File: src/cli/src/parser.c [hunk 2]
```diff
@@ -3,7 +3,95 @@
...
+    {
+        s+=strlen(delim);
```
Comment (ecoble) [inline src/cli/src/parser.c:32]: Same here with spaces around +=

File: src/cli/src/parser.c [hunk 3]
```diff
@@ -3,7 +3,95 @@
...
+        if (*s == '\0')
+        {
+            *save_ptr = s;
+            return NULL;
+            }
```
Comment (ecoble) [inline src/cli/src/parser.c:37]: Watch bracket spacing

File: src/cli/src/parser.c [hunk 4]
```diff
@@ -3,7 +3,95 @@
...
+    }
+    // Find the end of the token.
+    end = strstr (s, delim);
```
Comment (ecoble) [inline src/cli/src/parser.c:40]: No space between function name and parentheses

File: src/cli/src/parser.c [hunk 5]
```diff
@@ -3,7 +3,95 @@
...
+    tokenized_cmds *head = NULL;
+    char **save_ptr = &input;
+    
+    //puts tokenized segments of the inputted string in 
+    //the command line with a delimiter "AND" into a 
+    //utlist until the end of the string
+    char *token = strtokstr_r(input, "AND", save_ptr);
+    while(token != NULL)
+    {
+        tokenized_cmds *new_thing = malloc(sizeof(tokenized_cmds));
```
Comment (ecoble) [inline src/cli/src/parser.c:80]: Having a more descriptive name than "new_thing" would be good, maybe something like "new phrase"

**Review by ecoble — CHANGES_REQUESTED** [review]
This is looking great!

File: src/cli/src/cmd.c [hunk 6]
```diff
@@ -227,11 +227,13 @@ cmd *cmd_from_tokens(char **ts, lookup_t **table)
...
 }
 
+
 /* See cmd.h */
 cmd *cmd_from_string(char *s, chiventure_ctx_t *ctx)
 {
+    
     char **parsed_input = parse(s);
-    if(parsed_input == NULL)
+    if(s == NULL)
```
[the remark this answers is not in the document]
Comment (Beza20) [reply src/cli/src/cmd.c:236]: Yeah good catch. Amazing that it didn't make it crash. But we missed it when we adjusted the code while forgetting that after parsing with "and" we forgot to additionally parse with space.

File: src/cli/src/cmd.c [hunk 7]
```diff
@@ -227,9 +227,11 @@ cmd *cmd_from_tokens(char **ts, lookup_t **table)
     return output;
 }
 
+
```
Comment (ecoble) [inline src/cli/src/cmd.c:230]: Can we remove both of these extra new lines? So that we aren't unnecessarily changing this file.

File: src/cli/src/parser.c [hunk 8]
```diff
@@ -3,7 +3,95 @@
...
+        // Terminate the token and make *SAVE_PTR point past it.
+        memset(end, 0, strlen(delim));
+        *save_ptr = end + strlen(delim);
+        return s;
+}
+
+/* See parser.h */
+tokenized_cmds *parse_r(char *input)
+{
+    if(strcmp(input, "") == 0)
```
Comment (ecoble) [inline src/cli/src/parser.c:55]: Space after if

File: src/cli/src/parser.c [hunk 9]
```diff
@@ -3,7 +3,95 @@
...
+    {
+        return NULL;
+    }
+
+    //Changes the input to be all caps, for compatibility with commands/objects/directions
+    int i = 0;
+    char ch;
+
+    while(input[i])
```
Comment (ecoble) [inline src/cli/src/parser.c:64]: Space after while

File: src/cli/src/parser.c [hunk 10]
```diff
@@ -3,7 +3,95 @@
...
+    }
+
+    tokenized_cmds *head = NULL;
+    char **save_ptr = &input;
+    
+    //puts tokenized segments of the inputted string in 
+    //the command line with a delimiter "AND" into a 
+    //utlist until the end of the string
+    char *token = strtokstr_r(input, "AND", save_ptr);
+    while(token != NULL)
```
Comment (ecoble) [inline src/cli/src/parser.c:78]: Space after while

File: src/cli/src/parser.c [hunk 11]
```diff
@@ -3,7 +3,95 @@
...
+        tokenized_cmds *added_cmd = malloc(sizeof(tokenized_cmds));
+        added_cmd->cmds = token;
+        added_cmd->next = NULL;
+        LL_APPEND(head, added_cmd);
+        token = strtokstr_r(input, "AND", save_ptr);
+    }
+   
+    //If there are more than 4 words, parser returns NULL and does not attempt
+    //to pass the first four words as tokens
+    if(token != NULL)
```
Comment (ecoble) [inline src/cli/src/parser.c:89]: Space after if

File: src/ui/src/print_functions.c [hunk 12]
```diff
@@ -176,16 +177,25 @@ void print_cli(chiventure_ctx_t *ctx, window_t *win, int *quitval)
...
+
     if (ctx->game->mode->curr_mode == NORMAL) 
     {
-        cmd *c = cmd_from_string(cmd_string, ctx);
-        if (!c)
-        {
-            print_to_cli(ctx, "Error: Malformed input (4 words max)");
-        }
-        else
+        //iteratively goes through each tokenized cmds
```
Comment (ecoble) [inline src/ui/src/print_functions.c:183]: It would be better to make this a block comment with this style:
/*
 *
 */

**Review by ecoble — CHANGES_REQUESTED** [review]
A couple more style things but it's looking great!

File: src/cli/src/parser.c [hunk 13]
```diff
@@ -3,7 +3,95 @@
...
+{
+    char *end;
+    if (s == NULL)
+    {
+        s = *save_ptr;
+    }
+    else if (s == NULL || *s == '\0')
+    {
+        *save_ptr = s;
+        return NULL;
```
Comment (samdonio) [inline src/cli/src/parser.c:28]: I'm a little confused on whats going on here, isn't save_ptr just a pointer to s, so setting s = *save_ptr the same as saying s = s? Also, why test if s == NULL in the else if, if s == NULL then this will never be run?

File: src/cli/src/parser.c [hunk 14]
```diff
@@ -3,7 +3,95 @@
...
+    }
+    while (strstr(s,delim) == s)
```
Comment (samdonio) [inline src/cli/src/parser.c:30]: This is very minor but just for style this could be changed to strstr(s, delim)

File: src/cli/src/parser.c [hunk 15]
```diff
@@ -3,7 +3,95 @@
...
+    //Changes the input to be all caps, for compatibility with commands/objects/directions
+    int i = 0;
+    char ch;
+
+    while (input[i])
+    {
+        ch = toupper(input[i]);
+        input[i] = ch;
+        i++;
+    }
```
Comment (samdonio) [inline src/cli/src/parser.c:69]: This may need to be looked at when/if case insensitivity gets implemented.

**Review by samdonio — COMMENTED** [review]
Overall looks good, just have some small comments

File: src/cli/src/parser.c (hunk 14 above)
Comment (Beza20) [reply src/cli/src/parser.c:30]: well, the while loop needs to run only as long as strstr outputs s.

File: src/cli/src/parser.c (hunk 15 above)
Comment (Beza20) [reply src/cli/src/parser.c:69]: well when case sensitivity gets implemented the only thing might be that we would have no need for it but for now I believe it's fine.

File: src/cli/src/parser.c (hunk 13 above)
Comment (Beza20) [reply src/cli/src/parser.c:28]: we also wanted to check if this is an empty line. The helper function runs through an iteration the same way the built-in C function strtok_r does and is basically translated line by line and we have made sure it passes all tests.

File: src/cli/src/parser.c (hunk 14 above)
Comment (Beza20) [reply src/cli/src/parser.c:30]: it's not enough that it is just not outputting NULL

**Review by ecoble — APPROVED** [review]
Great job!

**Review by samdonio — APPROVED** [review]
Good job working on a pretty major change!

**Comment by Beza20** [comment]
we have enabled a user to input multiple commands at once using a connecting word "and." This was done by using a helper function <strtokstr_r> that has very similar structure as the C built-in function <strtok_r.> It parses through a string and tokenizes a segment of that string whenever it finds the inputed delimiter of a string type. Then we use the helper function to read the cmd line and store the tokenized segment of the input into a utlist. Note: Each element in the utlist has to further be parsed breaking it up using a space. And then in the UI module, we made sure that all of the command segments put into the utlist are iteratively changed into a command struct by calling the <cmd_from_string> and executed. And with that this pull request has introduced the "and" feature to chiventure resolving issue #275.

**Comment by Beza20** [comment]
we have enabled a user to input multiple commands at once using a connecting word "and." This was done by using a helper function <strtokstr_r> that has very similar structure as the C built-in function <strtok_r.> It parses through a string and tokenizes a segment of that string whenever it finds the inputed delimiter of a string type. Then we use the helper function to read the cmd line and store the tokenized segment of the input into a utlist. Note: Each element in the utlist has to further be parsed breaking it up using a space. And then in the UI module, we made sure that all of the command segments put into the utlist are iteratively changed into a command struct by calling the <cmd_from_string> and executed. And with that this pull request has introduced the "and" feature to chiventure resolving issue #275.

**Comment by ecoble** [comment]
Issue Score: ✔️++

Comments:
Amazing work on this issue! This was definitely a tricky issue, and a huge improvement to the parser! And your documentation on this PR is great too!
